### ISSUE TITLE
TypeError: Cannot read property 'route' of undefined

### ISSUE THREAD

**Issue by techyrajeev** [airbnb/enzyme#883]
I am trying to test below code - 
But when I am calling html() or render() method I am getting error.
I am using react-router v4 and enzyme 2.6.0

```
import React from 'react';
import { expect } from 'chai';
import { shallow, mount, render } from 'enzyme';
import { NavLink } from 'react-router-dom';

describe("<NavLink  />", () => {
    it("contains correct passed prop", () => {
        const comp = (
            <NavLink to="/home"  className={"test"}>
                NaveLink Test
            </NavLink>
        );
        const wrapper
            = shallow( comp );
        console.log(wrapper.html());
    });
});
```

Error : 

```
 at Route.computeMatch (node_modules/react-router-dom/node_modules/react-router/Route.js:66:22)
      at new Route (node_modules/react-router-dom/node_modules/react-router/Route.js:43:20)
      at node_modules/react-dom/lib/ReactCompositeComponent.js:295:18
      at measureLifeCyclePerf (node_modules/react-dom/lib/ReactCompositeComponent.js:75:12)
      at ReactCompositeComponentWrapper._constructComponentWithoutOwner (node_modules/react-dom/lib/ReactCompositeComponent.js:294:16)
      at ReactCompositeComponentWrapper._constructComponent (node_modules/react-dom/lib/ReactCompositeComponent.js:280:21)
      at ReactCompositeComponentWrapper.mountComponent (node_modules/react-dom/lib/ReactCompositeComponent.js:188:21)
      at Object.mountComponent (node_modules/react-dom/lib/ReactReconciler.js:46:35)
      at node_modules/react-dom/lib/ReactServerRendering.js:45:36
      at ReactServerRenderingTransaction.perform (node_modules/react-dom/lib/Transaction.js:140:20)
      at renderToStringImpl (node_modules/react-dom/lib/ReactServerRendering.js:43:24)
      at renderToStaticMarkup (node_modules/react-dom/lib/ReactServerRendering.js:83:10)
      at ShallowWrapper.<anonymous> (node_modules/enzyme/build/ShallowWrapper.js:765:88)
      at ShallowWrapper.single (node_modules/enzyme/build/ShallowWrapper.js:1504:25)
      at ShallowWrapper.html (node_modules/enzyme/build/ShallowWrapper.js:764:21)
      at Context.<anonymous> (test/components/navlink.spec.js:16:16)
```

**Comment by ljharb** [comment]
What's your component code look like?

**Comment by techyrajeev** [comment]
My component is already there in test case. Actually I am using NavLink from react-router-dom.

**Comment by ljharb** [comment]
Ah, sorry, I missed that. Calling `.html()` or `.render()` fully renders the entire `NavLink` tree - if you render your example component directly on the page, with nothing else, I'd expect you'd get the same error. Do you?

**Comment by techyrajeev** [comment]
Yes you are right.

**Comment by techyrajeev** [comment]
Wait. I did not get your answer. Please explain a little bit why that error is occurring?

**Comment by ljharb** [comment]
That component is required to be used inside a certain context from react-router. It cannot be completely rendered in isolation.  You may be able to provide a mock context that imitates the one that react-router provides, but otherwise I would stick to some assertions about what NavLink shallow-renders.

Separately, you should only ever test only components. NavLink is not yours; you do not need to test it.

**Comment by techyrajeev** [comment]
But I felt that each react component is like a well isolated system in itself so it can be tested in isolation as long as I am passing props and State required for that component. Also while instantiating NavLink did not complained anything​ about context related thing. 
I don't know from where route.js is coming​ into picture. May be I am missing some important stuff.

**Comment by ljharb** [comment]
That's not accurate; there's also context, but either way, you shouldn't be writing tests for code you didn't write yourself.

**Comment by techyrajeev** [comment]
Ok so context could be the issue. But regarding the test case What if I used NavLink to compose another component? I was doing that only, I wrote test case for my component only. But there also I was getting exactly same error. Why do I  bother about re testing react-router NavLink?  An example could be a header bar where NAV link can be used. 
So if I pickup somebody else component how do I know that there is hidden context as well other than props or state.
